### PR TITLE
Clean addresses with peer id before adding to addrbook

### PR DESCRIPTION
**Pull request by sukunrt**
Fixes: https://github.com/libp2p/go-libp2p/issues/2001

### CODE REVIEW

**Comment by sukunrt** [comment]
I'm cleaning the address before adding it to addressbook. The other option is to change the check in all transports. This seemed cleaner to me.

File: p2p/host/peerstore/test/addr_book_suite.go
Comment (MarcoPolo) [inline p2p/host/peerstore/test/addr_book_suite.go:151]: We should have a case that fails if the peer id in the multiaddr is != to the peer id passed in.

Comment (MarcoPolo) [reply p2p/host/peerstore/test/addr_book_suite.go:151]: The annoying thing is that `AddAddr` doesn't return an error. So the best we can do here is to fail silently and log an error in the implementations.

File: p2p/host/peerstore/pstoreds/addr_book.go
Comment (MarcoPolo) [inline p2p/host/peerstore/pstoreds/addr_book.go:601]: This should now take a peer id as well, and filter out multiaddrs that have an id component but don't match (and log an error when this happens)

File: p2p/host/peerstore/pstoremem/addr_book.go
Comment (MarcoPolo) [inline p2p/host/peerstore/pstoremem/addr_book.go:242]: same comment as above about the peer id check. Also I think it's okay to shadow the `addr variable` with the new multiaddr.

File: p2p/host/peerstore/pstoremem/addr_book.go
Comment (MarcoPolo) [inline p2p/host/peerstore/pstoremem/addr_book.go:288]: same comment as above.

**Review by MarcoPolo — CHANGES_REQUESTED** [review]
Thanks for picking this up! This is really helpful 🎉. There's a couple of things we should do before merging this:

* The changes mentioned in the review.
* We should update the implementations of [`Host.Connect`](https://pkg.go.dev/github.com/libp2p/go-libp2p/core/host#Host) so that they check that the multiaddrs with a `p2p` component in the `Addrs` field match the `ID` field. Filter out any mismatching multiaddrs (maybe log a warning). And return an error if no addresses are dialable because of mismatched peer ids. As well as removing the `p2p` component of the multiaddr before passing it down the call stack.

**Comment by sukunrt** [comment]
> * We should update the implementations of [`Host.Connect`](https://pkg.go.dev/github.com/libp2p/go-libp2p/core/host#Host) so that they check that the multiaddrs with a `p2p` component in the `Addrs` field match the `ID` field. Filter out any mismatching multiaddrs (maybe log a warning). And return an error if no addresses are dialable because of mismatched peer ids. As well as removing the `p2p` component of the multiaddr before passing it down the call stack.

All Host.Connect implementations use PeerStore.AddAddrs which will filter out for invalid ID, should I do it in the connect implementation as well? 
I was thinking of doing this:
```
func (h *BasicHost) Connect(ctx context.Context, pi peer.AddrInfo) error {
	// absorb addresses into peerstore
	h.Peerstore().AddAddrs(pi.ID, pi.Addrs, peerstore.TempAddrTTL)
    addrs := h.Peerstore().Addrs(pi.ID)
	if len(addrs) == 0 {
		return fmt.Errorf("no valid address found for peer %s", pi.ID)
	}

```

File: p2p/host/peerstore/test/addr_book_suite.go
Comment (sukunrt) [reply p2p/host/peerstore/test/addr_book_suite.go:151]: Added another test case with filtering on invalid peerId

File: p2p/host/peerstore/pstoreds/addr_book.go
Comment (sukunrt) [reply p2p/host/peerstore/pstoreds/addr_book.go]: fixed now

File: p2p/host/peerstore/pstoremem/addr_book.go
Comment (sukunrt) [reply p2p/host/peerstore/pstoremem/addr_book.go]: This should be fixed

File: p2p/host/peerstore/pstoremem/addr_book.go
Comment (sukunrt) [reply p2p/host/peerstore/pstoremem/addr_book.go]: fixed now.

**Comment by sukunrt** [comment]
@MarcoPolo Addressed review comments. Do the changes in Host.Connect look okay?

File: p2p/host/peerstore/pstoreds/addr_book.go
Comment (sukunrt) [reply p2p/host/peerstore/pstoreds/addr_book.go]: I'm logging a warning which was previously logged for nil addresses, should I make it an error?

**Comment by MarcoPolo** [comment]
> All Host.Connect implementations use PeerStore.AddAddrs which will filter out for invalid ID and log the warning, 

Good point. If there are no available addresses we'll still get that error from `Connect`. I think that's okay.

File: p2p/host/basic/basic_host.go
Comment (MarcoPolo) [inline p2p/host/basic/basic_host.go]: On second thought we already return a similar error here https://github.com/libp2p/go-libp2p/blob/master/p2p/net/swarm/swarm_dial.go#L340. So this change is not required. Sorry I added confusion here.

**Review by MarcoPolo — APPROVED** [review]
Some small nits. I pushed a change. Thanks a ton for fixing this :)

**Comment by MarcoPolo** [comment]
Will merge after CI passes.

File: p2p/host/basic/basic_host.go
Comment (sukunrt) [reply p2p/host/basic/basic_host.go]: that's fair. Thanks for fixing this!

**Comment by masih** [comment]
Thanking you folks for getting this issue resolved so quickly 🙇

**Pull request merged**
